apex/contrib/csrc/optimizers/multi_tensor_distopt_adam_kernel.cu

```diff
@@ -194,6 +194,170 @@ struct DistAdamFunctor
   }
 };
 
+/* Multi-tensor Adam with CUDA Graph Support
+ *
+ * Updates params in-place and outputs a copy with a desired datatype.
+ */
+template <typename T, typename GRAD_T, typename PARAM_OUT_T>
+struct DistAdamCapturableFunctor
+{
+  // Vectorized local compute
+  __device__ __forceinline__ static void local_step(
+    T p[ILP],
+    T m[ILP],
+    T v[ILP],
+    const GRAD_T g[ILP],
+    const float grad_scale,
+    const float beta1,
+    const float beta2,
+    const float beta1_correction,
+    const float beta2_correction,
+    const float eps,
+    const float lr,
+    adamMode_t mode,
+    const float weight_decay) {
+    if (mode == ADAM_MODE_0) { // L2
+#pragma unroll
+      for (int ii = 0; ii < ILP; ii++) {
+        float scaled_grad = (g[ii] * grad_scale) + (weight_decay * p[ii]);
+        float next_m = lerp(beta1, scaled_grad, m[ii]);
+        float next_v = lerp(beta2, scaled_grad*scaled_grad, v[ii]);
+        float next_m_unbiased = next_m / beta1_correction;
+        float next_v_unbiased = next_v / beta2_correction;
+        float denom = sqrtf(next_v_unbiased) + eps;
+        float update = next_m_unbiased / denom;
+        m[ii] = next_m;
+        v[ii] = next_v;
+        p[ii] -= lr * update;
+      }
+    } else { // weight decay
+#pragma unroll
+      for (int ii = 0; ii < ILP; ii++) {
+        float scaled_grad = g[ii] * grad_scale;
+        float next_m = lerp(beta1, scaled_grad, m[ii]);
+        float next_v = lerp(beta2, scaled_grad*scaled_grad, v[ii]);
+        float next_m_unbiased = next_m / beta1_correction;
+        float next_v_unbiased = next_v / beta2_correction;
+        float denom = sqrtf(next_v_unbiased) + eps;
+        float update = (next_m_unbiased / denom) + (weight_decay * p[ii]);
+        m[ii] = next_m;
+        v[ii] = next_v;
+        p[ii] -= lr * update;
+      }
+    }
+  }
+
+  __device__ __forceinline__ void operator()(
+    int chunk_size,
+    volatile int* noop_gmem,
+    TensorListMetadata<5>& tl,
+    const float* grad_scale_ptr,
+    const float beta1,
+    const float beta2,
+    const int* step,
+    const int bias_correction,
+    const float eps,
+    const float* lr,
+    adamMode_t mode,
+    const float weight_decay) const
+  {
+    if(*noop_gmem == 1)
+      return;
+
+    float beta1_correction = 1.0f, beta2_correction = 1.0f;
+    if (bias_correction == 1) {
+      beta1_correction = 1 - pow(beta1, *step);
+      beta2_correction = 1 - pow(beta2, *step);
+    }
+
+    int tensor_loc = tl.block_to_tensor[blockIdx.x];
+    int chunk_idx = tl.block_to_chunk[blockIdx.x];
+    int n = tl.sizes[tensor_loc];
+
+    const float grad_scale = *grad_scale_ptr;
```

Comment:
nit: might better to check whether or not `grad_scale_ptr` is `nullptr`

Reply:
Added some assertion at the beginning.